### PR TITLE
Filecatalog speedups

### PR DESCRIPTION
**Pull request by heckflosse**
#5375

### CODE REVIEW

File: rtgui/filecatalog.cc [hunk 1]
```diff
@@ -563,44 +566,44 @@ void FileCatalog::closeDir ()
...
                 }
 
-                const auto fext = fname.substr (lastdot + 1).lowercase ();
-                if (extensions.count (fext) == 0) {
+                if (extensions.count(fname.substr(lastdot + 1).lowercase()) == 0) {
                     continue;
                 }
 
-                names.emplace_back (Glib::build_filename (selectedDirectory, fname));
+                names.emplace_back(Glib::build_filename(selectedDirectory, fname));
```
Comment (Floessie) [inline rtgui/filecatalog.cc]: Don't think, `emplace_back()` helps here...

File: rtgui/filecatalog.cc [hunk 2]
```diff
@@ -1704,17 +1707,15 @@ void FileCatalog::reparseDirectory ()
...
-        bool found = false;
-
-        for (size_t j = 0; j < fileNameList.size(); j++)
-            if (nfileNameList[i] == fileNameList[j]) {
-                found = true;
-                break;
-            }
+    // build a set of collate-keys for faster search
+    std::set<std::string> oldNames;
+    for (size_t j = 0; j < fileNameList.size(); j++) {
```
Comment (Floessie) [inline rtgui/filecatalog.cc]: RB `for` loop?

File: rtgui/filecatalog.cc [hunk 3]
```diff
@@ -1704,17 +1707,15 @@ void FileCatalog::reparseDirectory ()
...
-    for (size_t i = 0; i < nfileNameList.size(); i++) {
-        bool found = false;
-
-        for (size_t j = 0; j < fileNameList.size(); j++)
-            if (nfileNameList[i] == fileNameList[j]) {
-                found = true;
-                break;
-            }
+    // build a set of collate-keys for faster search
+    std::set<std::string> oldNames;
```
Comment (Floessie) [inline rtgui/filecatalog.cc:1713]: Have you considered `std::unordered_set<>`? Would be interesting, if it's faster for your workload...

File: rtgui/filecatalog.cc [hunk 4]
```diff
@@ -1704,17 +1707,15 @@ void FileCatalog::reparseDirectory ()
...
+    for (size_t j = 0; j < fileNameList.size(); j++) {
+        oldNames.insert(fileNameList[j].collate_key());
+    }
 
-        if (!found) {
-            checkAndAddFile (Gio::File::create_for_parse_name (nfileNameList[i]));
+    for (size_t i = 0; i < nfileNameList.size(); i++) {
+        if (oldNames.count(nfileNameList[i].collate_key()) == 0) {
```
Comment (Floessie) [inline rtgui/filecatalog.cc]: I'd go for `find()` here... And a RBFL for `nfileNameList`.

File: rtgui/filecatalog.cc (hunk 3 above)
Comment (heckflosse) [reply rtgui/filecatalog.cc:1713]: I just tested using an unordered_set. No difference in processing time. I guess the most expensive part here still are the collate_key() calls.

**Comment by heckflosse** [comment]
Last commit speeds up the following case (not only this case, but especially):

1) Open a folder with only a few files and apply a find-filter
2) then open a folder with a large amount of files without resetting the find-filter before

**Comment by heckflosse** [comment]
Imho the pr has reached a state where it can be reviewed for a merge.

@Floessie @Beep6581 @Hombre57 Please test and review if time permits

**Comment by Beep6581** [comment]
I created a folder with 5000 JPEGs on an SSD drive using the EXT4 filesystem in Linux.

I opened the folder using `dev` `5.6-184-g`2dee88428, took a few minutes to generate the cache. I navigated to a different folder, and then back to this one. Re-opening it took about 1 minute for the "File Browser [5000]" counter to reach 5000, at which point I saw filenames but had to wait another minute for thumbnails to appear. Deleting one of the files also froze RT for about 2 minutes. I went to a different folder, then went back to this one and waited for everything to settle, then using the "Find" filename filter is instant (I searched for "123"). Clearing the filter is instant too. Clicking the "Find" textbox the first time took a few seconds before the flashing cursor appeared.

I opened the folder using `filecatalog_speedups` `5.6-186-g`d9c93e77a. It took ~5 seconds for the file counter to show 5000 and for filenames to appear, and another 30 seconds for thumbnails to appear. Clicking the "Find" textbox also took a few seconds for the flashing cursor to appear. Finding "123" was instant. Clearing was instant. Deleting a file was instant. I searched for "123", went to a different folder, and went back to this one - this time loading all thumbs only took about 6 seconds.

**Comment by heckflosse** [comment]
@Beep6581 Thanks for testing :+1: . I would like to merge the pr to get more tests. Any objections?

**Comment by Hombre57** [comment]
No objection :+1:

**Pull request merged**